src/semidiscretization/semidiscretization_euler_plasma.jl

```diff
@@ -0,0 +1,539 @@
+    @unpack semi_euler, semi_plasma, parameters, plasma_counter, cache = semi
+
+    u_euler = wrap_array(u_ode, semi_euler)
+    u_plasma = wrap_array(cache.u_ode, semi_plasma)
+    du_plasma = wrap_array(cache.du_ode, semi_plasma)
+
+    # set up main loop
+    finalstep = false
+    @unpack n_iterations_max, cfl, resid_tol, timestep_plasma = parameters
+    iter = 0
+    tau = zero(real(semi_plasma.solver)) # Pseudo-time
+
+    # iterate plasma solver until convergence or maximum number of iterations are reached
+    @unpack equations = semi_plasma
+    while !finalstep
+        dtau = @trixi_timeit timer() "calculate dtau" begin
+            cfl * max_dt(u_plasma, tau, semi_plasma.mesh,
+                   have_constant_speed(equations), equations,
+                   semi_plasma.solver, semi_plasma.cache)
+        end
+
+        # evolve solution by one pseudo-time step
+        time_start = time_ns()
+        timestep_plasma(cache, u_euler, tau, dtau, parameters, semi_plasma)
+        runtime = time_ns() - time_start
+        put!(plasma_counter, runtime)
+
+        # update iteration counter
+        iter += 1
+        tau += dtau
+
+        # check if we reached the maximum number of iterations
+        if n_iterations_max > 0 && iter >= n_iterations_max
+            @warn "Max iterations reached: Plasma solver failed to converge!" residual=maximum(abs,
+                                                                                               @views du_plasma[1,
+                                                                                                                ..,
+                                                                                                                :]) tau=tau dtau=dtau
+            finalstep = true
+        end
+
+        # this is an absolute tolerance check
+        if maximum(abs, @views du_plasma[1, .., :]) <= resid_tol
+            finalstep = true
+        end
+    end
+
+    return nothing
+end
+
+# Integrate plasma solver for 2N-type low-storage schemes
+function timestep_plasma_2N!(cache, u_euler, tau, dtau, plasma_parameters,
+                             semi_plasma,
+                             a, b, c)
+    electric_scale = 1 /
+                     (parameters.scaled_debye_length * parameters.scaled_debye_length)
+
+    # Note that `u_ode` is `u_plasma` in `rhs!` above
+    @unpack u_ode, du_ode, u_tmp1_ode = cache
+    u_tmp1_ode .= zero(eltype(u_tmp1_ode))
+    du_plasma = wrap_array(du_ode, semi_plasma)
+    for stage in eachindex(c)
+        tau_stage = tau + dtau * c[stage]
+
+        # rhs! has the source term for the harmonic problem
+        # We don't need a `@trixi_timeit timer() "rhs!"` here since that's already
+        # included in the `rhs!` call.
+        rhs!(du_ode, u_ode, semi_plasma, tau_stage)
+
+        # TODO: Find a way to pass dimensions of semi_euler so that we can replace 4 with 3 + ndims(semi_euler)
```

Comment:
One way to implement this would be `ndims(semi_plasma)` where you might need to add the `ndims` function as here:

https://github.com/trixi-framework/Trixi.jl/blob/ce47aeaf9417ef687aefda848d53a36938a566c3/src/semidiscretization/semidiscretization_euler_gravity.jl#L218